### ISSUE TITLE
Mockingboard configuration

### ISSUE THREAD

**Issue by roughana** [a2-4am/4cade#43]
There are various lists of titles that support the Mockingboard. e.g.
https://atariage.com/forums/topic/228710-games-with-mockingboard-support/

I've got a Mockingboard clone in my Platinum //e in slot 5. Moving it to slot 4 on my machine is difficult because of the positioning of the jacks on the card and the restricted backplane that the Platinum //e has. It just won't go in slot 4 without physically modifying something. It was difficult enough to make it fit in slot 5. If I did move it to 4 then where would I put the mouse card?

Feature request: Configuration/detection of Mockingboard in Total Replay and patching of Mockingboard capable games to use the configured/detected slot.

I'd be willing to help with the endeavour.

**Comment by roughana** [comment]
After more understanding, I gather that most of the titles in that list will apparently auto-configure for the slot that the Mockingboard is found in. However, this did not work for me with my configuration. Not sure if that is a problem with my Super Mockingboard II Clone in the Platinum //e International I have or something else. Looking or confirmation of what works from others.
Apparently Zaxxon only has support in a specific version.

**Comment by a2-4am** [comment]
IIRC, the version of Zaxxon in Total Replay is an earlier version that does not support the Mockingboard at all. Support for the later version is on our TODO list.

**Comment by peterferrie** [comment]
Yes, the existing Zaxxon in Total Replay doesn't support the
Mockingboard, but we have a .dsk of the version that does.

**Comment by roughana** [comment]
Does the version of Lancaster support the Mockingboard? There is a Wildcard capture of a version that does. I've reached out towards the owner to obtain an image (Passport/AppleSauce) of the whole disk so that it can be preserved properly.

**Comment by peterferrie** [comment]
It does, but it's hard-coded to slot 4.

**Comment by roughana** [comment]
I'm working on a Mockingboard detection routine. It is based on the Mockingboard Toolkit SLOTFINDER with optimisation from https://github.com/deater/dos33fsprogs/blob/master/mode7_demo/mockingboard.s

**Comment by roughana** [comment]
Proposal:
init keeps the slot somewhere
games.conf is updated with another flag for each game: 0 for no mockingboard change needed, 1 for mockingboard change needed - only games that do not detect the slot automatically need to be set with 1
when mockingboard has been detected and game is selected that has 1 set, and slot is not slot 4 (most games without config assume slot 4) then patch before control is passed to the game.
patch could be a file along side game that has locations for where each patch needs to be applied e.g.
byte offsets for storing $Cx
alternatively, each game could have multiple versions for each slot with patches already applied and the loader picks the right one based on slot

**Comment by a2-4am** [comment]
I like the idea of Mockingboard auto-configuration, and we will definitely look into it in a future version. I suspect that the new prelaunch architecture in v3 could be used for per-game dynamic patching, in lieu of another games.conf flag.

In the meantime, I will be closing your untested pull request.

**Comment by roughana** [comment]
I've just explored prelauncher and it looks perfect for the patching. How much program space is available at the prelaunch load location? Seems to start at $106? So if I don't need the stack... 
I need to determine if there is room for the mockingboard detection routine (~$2A bytes) or whether it will need to be added as a common routine and call it like enable_accel/disable_accel.
I'll work on a prelauncher for Berzap as a POC.

**Comment by roughana** [comment]
Hmmm, maybe not. That needs some more knowledge about the specific PO crack to be able to decompress etc.
Here are the offsets within BZ.MAIN where $C4xx is used:
3fcd
3fd5
3fe1
3fe6
3feb
3ff0

**Comment by peterferrie** [comment]
We can do a single detection during the init phase.  The card isn't
going to be moved after start-up.
Then yes, it's a per-game patching via callback after decompression,
just like the cheat mechanism.
Thanks for the list.

**Comment by roughana** [comment]
I’ve started crafting a call to detect Mockingboard and store slot from 4cade.init. As the slot needs to be stored I see no immediate value in also taking a bit from machine status. I’m targetting FFF8 for the slot. The detection relies on a cycle count so it needs to be done after disableaccel is available. I’m almost at stage of being able to test. I think Zaxxon could be a good candidate as it has prelauncher already.

**Comment by roughana** [comment]
I believe this is the set of Total Replay titles that need patching:
Berzap is not a good candidate yet because it does not yet have a prelauncher.
Lady Tut is not a good candidate as the available patch location is zero page $F0-$FF - not big enough.
Lancaster could be a good candidate if it needs to be patched (I've had conflicting reports about it assuming slot 4 and need to confirm)
Pitfall II is a good candidate as it is smaller size.
Zaxxon is not a good candidate yet because doesn't appear to have been updated with Mockingboard capable version yet.

**Comment by peterferrie** [comment]
Lady Tut could just patch $AC-AD with slot and type and then patch $84B2 to jump to $854B to bypass the questions.

**Comment by roughana** [comment]
I have a tested solution for Lancaster. I'll submit the code changes tomorrow.

**Comment by roughana** [comment]
Pull request available
https://github.com/a2-4am/4cade/pull/62/

**Comment by roughana** [comment]
Berzap has lots of mockingboard code. The stack is probably not large enough for the patch routine.
The offsets above correspond to $1FCD-1FF0
Then there's loads more $9743-990C & $AD0B-AD62.
Potentially looping through 4x page lookups & replacements with a subroutine called 4 times and adjust the page base between calls and possibly the offset to start from as well.
The good news is that $C4 as an opcode is CPY ZP which is rarely used and not used within these ranges. There is no usage of $C4 as a low address byte or as a branch.

**Comment by roughana** [comment]
Another alternative is a lookup table which may be more generic and allow re-use with other prelaunchers but would still not fit in stack space.
Potentially a table lookup routine and patcher could be provided in the launcher and then the prelauncher calls that with the table address. Still need a mechanism to load in the table though.

**Comment by roughana** [comment]
I have a tested solution for Pitfall II.
Prelauncher calls $6000 to start the game.
$612A moves code from pages $2E,$2F to $AE,$AF for execution.
Therefore patches are in pages $2E and $2F

**Comment by roughana** [comment]
I created a patch for Berzap just to see how big it was. It fits in one page. Yay. Now to test it.

**Comment by roughana** [comment]
I have a tested solution for Berzap! This has been added to existing Pull Request
The original prelauncher left the patch in place and the mainline code calls it frequently.
As the patch is now so long, it will not survive on stack. I have removed the patch from the mainline code so it is only run once.
I tried to determine where the configuration graphic was so that it could be changed ("mockingboards must be in slot 4") but I have not found this yet - it doesn't appear to be in memory as a whole screen. Any help with this would be appreciated.
Warning: pressing ctrl-reset to return to Total Replay while the music is playing leaves a note playing without end on my clone card. Enter the config screen first will stop the music.
Future patch could potentially adjust reset vector to stop music.
